ENH: improve type flexibility for parallel reduction "iterator" type

Split off from https://github.com/kokkos/pykokkos/pull/17/files#r852448609 to break up the engineering tasks a bit.

In this code example:

```python
@pk.workload
class SquareSumInt:
    def __init__(self, n):
        self.N: int = n
        self.total: pk.int64 = 0

    @pk.main
    def run(self):
        self.total = pk.parallel_reduce(self.N, self.squaresum)

    @pk.workunit
    def squaresum(self, i: int, acc: pk.Acc[pk.int64]):
        acc += i * i

```

The `i` in `squaresum` appears to not tolerate being set as a `pk.int64`, with the monster traceback below the fold.

<details>

```
========================================================================================================================================================= FAILURES =========================================================================================================================================================
______________________________________________________________________________________________________________________________________________ test_squaresum_types[int64-10] ______________________________________________________________________________________________________________________________________________

self = <pykokkos.core.translators.static.StaticTranslator object at 0x14769810b0d0>, source = (['class SquareSumInt:\n', '    def __init__(self, n):\n', '        self.N: int = n\n', '        self.total: pk.int64 = 0\n', '\n', '    @pk.main\n', ...], 62)

    def translate_workunits(self, source: Tuple[List[str], int]) -> Tuple[Dict[cppast.DeclRefExpr, Tuple[str, cppast.MethodDecl]], bool]:
        """
        Translate the workunits
    
        :param source: the python source code of the workload
        :returns: a tuple of a dictionary mapping from workload name
            to a tuple of operation name and source, and a boolean
            indicating whether the workunit has a call to pk.rand()
        """
    
        node_visitor = WorkunitVisitor(
            {}, source, self.pk_members.views, self.pk_members.pk_workunits,
            self.pk_members.fields, self.pk_members.pk_functions,
            self.pk_members.classtype_methods, self.pk_import, debug=True)
    
        workunits: Dict[cppast.DeclRefExpr, Tuple[str, cppast.MethodDecl]] = {}
    
        has_rand_call: bool = False
        for n, w in self.pk_members.pk_workunits.items():
            try:
>               workunits[n] = node_visitor.visit(w)

pykokkos/core/translators/static.py:199: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pykokkos.core.visitors.workunit_visitor.WorkunitVisitor object at 0x14769810b640>, node = <_ast.FunctionDef object at 0x14774246fb20>

    def visit(self, node):
        """Visit a node."""
        method = 'visit_' + node.__class__.__name__
        visitor = getattr(self, method, self.generic_visit)
>       return visitor(node)

../spack/opt/spack/linux-centos8-cascadelake/gcc-9.4.0/python-3.8.12-v3zf3eymsvk5kys3vynoj7xljq4ensj2/lib/python3.8/ast.py:371: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pykokkos.core.visitors.workunit_visitor.WorkunitVisitor object at 0x14769810b640>, node = <_ast.FunctionDef object at 0x14774246fb20>

    def visit_FunctionDef(self, node: ast.FunctionDef) -> Union[str, Tuple[str, cppast.MethodDecl]]:
        if self.is_nested_call(node):
            params: List[cppast.ParmVarDecl] = [a for a in self.visit(node.args)]
            body = cppast.CompoundStmt([self.visit(b) for b in node.body])
    
            workunit = cppast.LambdaExpr("[&]", params, body)
            self.nested_work_units[node.name] = workunit
    
            return ""
    
        else:
            operation: Optional[str] = self.get_operation_type(node)
            if operation is None:
                self.error(node.args, "Incorrect types in workunit definition")
    
            tag_type = cppast.ClassType(f"const {node.name}")
            tag_type.is_reference = True
            tag = cppast.ParmVarDecl(tag_type, cppast.DeclRefExpr(""))
    
            params: List[cppast.ParmVarDecl] = [tag]
>           params.extend(self.visit(node.args))

pykokkos/core/visitors/workunit_visitor.py:43: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pykokkos.core.visitors.workunit_visitor.WorkunitVisitor object at 0x14769810b640>, node = <_ast.arguments object at 0x14774246fb50>

    def visit(self, node):
        """Visit a node."""
        method = 'visit_' + node.__class__.__name__
        visitor = getattr(self, method, self.generic_visit)
>       return visitor(node)

../spack/opt/spack/linux-centos8-cascadelake/gcc-9.4.0/python-3.8.12-v3zf3eymsvk5kys3vynoj7xljq4ensj2/lib/python3.8/ast.py:371: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pykokkos.core.visitors.workunit_visitor.WorkunitVisitor object at 0x14769810b640>, node = <_ast.arguments object at 0x14774246fb50>

    def visit_arguments(self, node: ast.arguments) -> List[cppast.ParmVarDecl]:
        args: List[ast.arg] = node.args
    
        if len(args) == 0:
            return []
    
        # If the first argument does not have an annotation, then this is a nested workunit
        is_nested: bool = True if args[0].annotation is not None else False
    
        self_arg: ast.arg = args[0]
        if not is_nested and self_arg.arg != "self":
            self.error(args[0], "First argument has to be \"self\"")
    
        # Skip self argument
        if not is_nested:
            args = args[1:]
    
        cpp_args: List[cppast.ParmVarDecl] = []
    
        # Visit all tid args, could be more than one for MDRangePolicies.
        # Stop when the accumulator is reached or there are no more args.
        for a in args:
            is_acc: bool = isinstance(a.annotation, ast.Subscript)
            if is_acc:
                break
>           cpp_args.append(self.visit_arg(a))

pykokkos/core/visitors/workunit_visitor.py:168: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pykokkos.core.visitors.workunit_visitor.WorkunitVisitor object at 0x14769810b640>, node = <_ast.arg object at 0x14774246fbb0>

    def visit_arg(self, node: ast.arg) -> cppast.ParmVarDecl:
        if node.annotation is None:
            self.error(node, "Missing type annotation")
    
        decltype: cppast.Type = visitors_util.get_type(node.annotation, self.pk_import)
        if decltype is None:
            self.error(node, "Type not supported")
    
        # If argument is pk.TeamMember (hierarchical parallelism)
        is_hierachical: bool = isinstance(node.annotation, ast.Attribute)
    
        if is_hierachical:
>           decltype.typename = f"const {decltype.typename}"
E           AttributeError: can't set attribute

pykokkos/core/visitors/workunit_visitor.py:203: AttributeError

During handling of the above exception, another exception occurred:

series_max = 10, dtype = <class 'numpy.int64'>

    @pytest.mark.parametrize("series_max", [10, 5000, 90000])
    @pytest.mark.parametrize("dtype", [np.float64, np.int64])
    def test_squaresum_types(series_max, dtype):
        # related to gh-6
    
        # Enforce pykokkos reduction return view type
        # matches accumulator type, indirectly via
        # consistency with NumPy for lack
        # of overflow, etc.
        expected = np.sum(np.arange(series_max, dtype=dtype) ** 2)
        if dtype == np.float64:
            ss_instance = SquareSumFloat(series_max)
        elif dtype == np.int64:
            ss_instance = SquareSumInt(series_max)
>       pk.execute(pk.ExecutionSpace.OpenMP, ss_instance)

tests/test_parallelreduce.py:91: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pykokkos/interface/parallel_dispatch.py:247: in execute
    runtime_singleton.runtime.run_workload(space, workload)
pykokkos/core/runtime.py:45: in run_workload
    members: Optional[PyKokkosMembers] = self.compiler.compile_object(module_setup, space, km.is_uvm_enabled())
pykokkos/core/compiler.py:146: in compile_object
    self.compile_entity(module_setup.main, module_setup, entity, parser.get_classtypes(), space, force_uvm, members)
pykokkos/core/compiler.py:201: in compile_entity
    functor, bindings = translator.translate(entity, classtypes)
pykokkos/core/translators/static.py:67: in translate
    workunits, has_rand_call = self.translate_workunits(source)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pykokkos.core.translators.static.StaticTranslator object at 0x14769810b0d0>, source = (['class SquareSumInt:\n', '    def __init__(self, n):\n', '        self.N: int = n\n', '        self.total: pk.int64 = 0\n', '\n', '    @pk.main\n', ...], 62)

    def translate_workunits(self, source: Tuple[List[str], int]) -> Tuple[Dict[cppast.DeclRefExpr, Tuple[str, cppast.MethodDecl]], bool]:
        """
        Translate the workunits
    
        :param source: the python source code of the workload
        :returns: a tuple of a dictionary mapping from workload name
            to a tuple of operation name and source, and a boolean
            indicating whether the workunit has a call to pk.rand()
        """
    
        node_visitor = WorkunitVisitor(
            {}, source, self.pk_members.views, self.pk_members.pk_workunits,
            self.pk_members.fields, self.pk_members.pk_functions,
            self.pk_members.classtype_methods, self.pk_import, debug=True)
    
        workunits: Dict[cppast.DeclRefExpr, Tuple[str, cppast.MethodDecl]] = {}
    
        has_rand_call: bool = False
        for n, w in self.pk_members.pk_workunits.items():
            try:
                workunits[n] = node_visitor.visit(w)
                has_rand_call = has_rand_call or node_visitor.has_rand_call
                if node_visitor.has_rand_call:
                    workunit: cppast.MethodDecl = workunits[n][1]
                    self.add_rand_pool_state(workunit)
                    node_visitor.has_rand_call = False
            except:
                print(f"Translation of {w} failed")
>               sys.exit(1)
E               SystemExit: 1

pykokkos/core/translators/static.py:207: SystemExit
--------------------------------------------------------------------------------------------------------------------------------------------------- Captured stdout call ---------------------------------------------------------------------------------------------------------------------------------------------------
Translation of <_ast.FunctionDef object at 0x14774246fb20> failed

```

</details>

My current plan is to label the test cases that overflow in gh-17 as a result of this limitation with an `xfail` (known failure) for now, and point to this issue.

--- a/pykokkos/core/visitors/workunit_visitor.py
+++ b/pykokkos/core/visitors/workunit_visitor.py
@@ -200,7 +200,11 @@ class WorkunitVisitor(PyKokkosVisitor):
         is_hierachical: bool = isinstance(node.annotation, ast.Attribute)
 
         if is_hierachical:
-            decltype.typename = f"const {decltype.typename}"
+            try:
+                decltype.typename = f"const {decltype.typename}"
+            except AttributeError:
+                decltype._type = f"const {decltype.typename.value}"
+
             decltype.is_reference = True
 
         declname = cppast.DeclRefExpr(node.arg)

--- a/tests/test_parallelreduce.py
+++ b/tests/test_parallelreduce.py
@@ -70,7 +70,7 @@ class SquareSumInt:
         self.total = pk.parallel_reduce(self.N, self.squaresum)
 
     @pk.workunit
-    def squaresum(self, i: int, acc: pk.Acc[pk.int64]):
+    def squaresum(self, i: pk.int64, acc: pk.Acc[pk.int64]):
         acc += i * i
 
 
@@ -79,8 +79,6 @@ class SquareSumInt:
 def test_squaresum_types(series_max, dtype):
     # check for the ability to match NumPy in
     # sum of squares reductions with various types
-    if series_max == 90000 and dtype == np.int64:
-        pytest.xfail("see gh-24")
     expected = np.sum(np.arange(series_max, dtype=dtype) ** 2)
     if dtype == np.float64:
         ss_instance = SquareSumFloat(series_max)

